### PR TITLE
Teal App Development blog post

### LINKED ISSUE #140: Blog Post:  How fast is it to make a teal app?
### Blog Post

Vague idea:

Teal makes it fast and easy to make simple RShiny Apps. But how fast, and how easy? How little code do we need to make a working teal app (e.g. with a demog table and ae table) which we can then customise with our study data)? 

The example code could use `{phamraverseadam}` data.

@khatril are you or others in your team interested in this?

### PR DESCRIPTION
**Pull request by gigikenneth**
Thank you for your Pull Request! We have developed this task checklist to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your blog post.

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #140 into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update), and make sure the corresponding issue is linked in the Development section on the right hand side
- [x] Run the script from `CICD.R` line by line to first check the spelling in your post and then to make sure your code is compatible with our code-style. Address any incongruences by following the instructions in the file! 
- [x] Choose (possibly several) `tag(s)` or categories from the current list: `c("Metadata", "SDTM", "ADaM", "TLG", "Shiny", "Community", "Conferences", "Submissions", "Technical")` for your blog post. If you cannot find anything that fits your blog post, add your own tag! We occasionally tidy up all `tags` for consistency.
- [x] Add a short description for your blog post in the `description` field at the top of the markdown document.
- [x] Blog post is *short*, *personalized*, *reproducible* and *readable*
- [x] Add a disclaimer at the top of your post, if appropriate (example: **Disclaimer** 
This blog contains opinions that are of the authors alone and do not necessarily reflect the strategy of their respective organizations.)
- [x] Address all merge conflicts and resolve appropriately
- [x] Assign two of us (@bms63, @manciniedoardo, @StefanThoma, @kaz462) as reviewers in the PR.
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+```{r setup, include=FALSE}
+long_slug <- "2024-07-12_teal_app_development_pharmaverseadam"
+# renv::use(lockfile = "renv.lock")
+```
+
+<!--------------- post begins here ----------------->
+
+Every developer must solve two difficult problems when creating a Shiny application (in fact, any application) from the ground up: software architecture and data design. In the world of clinical data analysis, however, much development has been aimed at providing a jump-start approach to creating R/Shiny applications that would take away most of the pain caused by these two problems.
+
+This blog should help you get an idea of how easy it is to get started with the [pharmaverse](https://pharmaverse.org/) ecosystem. We will create an interactive clinical data dashboard using `{teal}` and will use `{pharmaverseadam}` as the data source.
```
Comment (bms63) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:26]: can you link teal and pharmaverseadam - we have the links package to do this - see admiral 1.1.1 post

**Review by bms63 — CHANGES_REQUESTED** [review]
whoa - this is a very clear and scaffold article.  An amazing addition to the blog!!

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd (hunk 1 above)
Comment (gigikenneth) [reply posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd]: Done!

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+First, you would need to install the two packages. It is recommended to use [`{pak}`](https://pak.r-lib.org/) to take advantage of parallel downloads and builds (note that you can use it together with [`{renv}`](https://rstudio.github.io/renv/articles/renv.html) for locking dependencies).
+
+``` r
+pak::pkg_install(c("teal", "pharmaverseadam"))
+```
+
+## **Building a Simple Teal App**
+
+Next, we would create a simple app - basically following [the official teal guide](https://insightsengineering.github.io/teal/latest-tag/articles/getting-started-with-teal.html), but we will use datasets from the {pharmaverseadam} package.
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:36]:
```suggestion
Next, we would create a simple app - basically following [the official teal guide](https://insightsengineering.github.io/teal/latest-tag/articles/getting-started-with-teal.html), but we will use datasets from the `{pharmaverseadam}` package.
```

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  modules = example_module()
+)
+
+# Step 4: run shiny app
+shinyApp(app$ui, app$server)
+```
+
+### **Let’s Take a Closer Look at Each Step:**
+
+-   **At step 1**, we import libraries to use functions and datasets exported by those.
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:61]:
```suggestion
-   **In step 1**, we import libraries to use functions and datasets exported by those.
```
(just to align with other steps)

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+-   **In step 2**, we create a teal data object that can be used in a teal app. There are a couple of subtle details here: instead of the default teal_data() function, we use a special wrapper around it, designed specifically for clinical trial data - cdisc_data(). Its advantage is that it will automatically generate join_keys() for the datasets that we pass in. Please note, that the dataset names should be all caps. The resulting data object is an s4 class instance, and we can verify that the output is correct by checking data names and join_keys slots:
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:63]: Use backticks for functions here - and perhaps link to them as well using the `link` package if that works?

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+Now we have reproducible data. But what about the app itself? Surely, [`{teal}`](https://insightsengineering.github.io/teal/latest-tag/) features don’t end here. There is a collection of pre-built shiny modules that can be used in teal applications. We can install them with pak:
+
+``` r
+pak::pkg_install(
+  c("sparkline", "teal.modules.general", "teal.modules.clinical")
+)
+```
+
+First, I suggest exploring the “general” modules that are applicable to any kind of data. The only (!) thing we need to do, is to add two more modules to the app initializer:
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:118]: As the blog has multiple authors, suggest to use "we" throughout

```suggestion
First, we suggest exploring the “general” modules that are applicable to any kind of data. The only (!) thing we need to do, is to add two more modules to the app initializer:
```

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  )
+)
+```
+
+This is possible thanks to the magic that [`{teal}`](https://insightsengineering.github.io/teal/latest-tag/) is doing under the hood - passing the data object to each module. And now we will have access to a nice tabular view of the data, and a tool to explore each variable in greater detail.
+
+![](images/Asset 4.gif)\
+This is great on its own, but as a bonus, we even get the ability to build a report based on some of the modules that we have. For example, we can generate some plots in the variable browser, add them to the report and preview it. Some modules would also add a block of R code showing how to get the exact same data that was used to generate a report card.
+
+Finally, let’s add a simple barchart module that comes from the clinical modules package. We will use an example from the {teal.modules.clinical} documentation:
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:136]: Link package here and use backticks

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    barchart_module
+  )
+)
+
+shinyApp(app$ui, app$server)
+```
+
+## **Conclusion** 
+
+In conclusion, [Teal](https://github.com/insightsengineering/teal) and [PharmaverseAdam](https://github.com/pharmaverse/pharmaverseadam) make it much easier to create interactive and reproducible clinical data dashboards. By following this guide, you can quickly build a Shiny app that not only visualizes your data but also maintains reproducibility and customization options.
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:206]: `{pharmaverseadam}` should be all lowercase and within curly brackets - not sure if you want to match teal to that

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+)
+
+shinyApp(app$ui, app$server)
+```
+
+## **Conclusion** 
+
+In conclusion, [Teal](https://github.com/insightsengineering/teal) and [PharmaverseAdam](https://github.com/pharmaverse/pharmaverseadam) make it much easier to create interactive and reproducible clinical data dashboards. By following this guide, you can quickly build a Shiny app that not only visualizes your data but also maintains reproducibility and customization options. 
+
+Get the latest updates from the Pharmaverse delivered to your inbox. [Subscribe to our newsletter today.](https://forms.office.com/e/AJu4xWEGJx)
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:208]: "pharmaverse" is generally styled as lowercase in this blog

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+title: "Simplifying Clinical Data Dashboards with Teal and PharmaverseAdam"
+author:
+  - name: Pavel Demin
+  - name: APPSILON
+  - name: Dror Berel
+description: "Learn how to quickly develop interactive and reproducible clinical data dashboards using Teal and PharmaverseAdam." # <---- please write a description
+date: "2024-07-12"
+# please do not use any non-default categories.
+# You can find the default categories in the repository README.md
+categories: [Technical, Community]
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:11]: We also have a shiny tag - suggest to add

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 10]
```diff
@@ -0,0 +1,208 @@
+---
+title: "Simplifying Clinical Data Dashboards with Teal and PharmaverseAdam"
+author:
+  - name: Pavel Demin
+  - name: APPSILON
```
Comment (manciniedoardo) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:5]: Perhaps move appsilon to end of authors as it's a company? Just to align with other posts.

**Review by manciniedoardo — CHANGES_REQUESTED** [review]
😮 amazing post!! I love the gifs and it reads super well, congrats! Will be good to finally get a post on this topic out.

Left some comments 😄

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+  ADAE <- pharmaverseadam::adae
+  ADSL <- pharmaverseadam::adsl
+})
+datanames(data) <- c("ADAE", "ADSL")
+join_keys(data) <- default_cdisc_join_keys[datanames(data)]
+```
+
+The recommended method is to generate the data using a within function. However, this method requires manually providing databases using a helper function. We also need to provide `join_keys` ourselves, but given that the data names are standard ADaM names, we can take advantage of a special `default_cdisc_join_keys` object.
+
+It is also worth noting that [`{teal}`](https://insightsengineering.github.io/teal/latest-tag/) has its implementation of `within` generic.
```
Comment (StefanThoma) [inline posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd:104]: @bms63 should we suggest to use the {link} package, or leave it up to the author?

File: posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd (hunk 4 above)
Comment (gigikenneth) [reply posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd]: all done!

**Comment by bms63** [comment]
Hi @manciniedoardo I am out of office tomorrow and next week.  Can you merge in and announce please.  tyty

**Comment by manciniedoardo** [comment]
@bms63 sure - are we letting the spelling slide?

**Review by manciniedoardo — APPROVED** [review]

**Pull request merged**